### PR TITLE
settings page for voting duration

### LINKED ISSUE #116: Dev: Governance Setting – Vote Duration
See https://github.com/NEAR-DevHub/neardevhub-treasury-dashboard/issues/86 for Governance Settings background. 

# Background
Our Treasury App currently lacks a dedicated section for admins to configure governance settings (see [#86](https://github.com/NEAR-DevHub/neardevhub-treasury-dashboard/issues/86) for background). Specifically, there is no way for admins to adjust the "voting duration," or the number of days a vote is active before it expires. This limits the flexibility of governance processes, especially when there’s a need to adapt voting periods to different requirements

This ticket focuses on implementing a "Voting Duration" component within the Settings page, which would allow admins to configure the number of days a vote is open for proposals before it expires.

# User Story
As an admin (user with permissions to manage members & setting), I want the ability to configure the voting duration so that I can choose the number of days a vote remain active before it expires, allowing me to manage voting timelines that work for my team.

# Acceptance Criteria 

## 1. Voting Duration Configuration Section
- [x] A new "Voting Duration" section is added to the Treasury settings
  - [x] Title: "Voting Duration"
  - [x] Instructions text: "Set the number of days a vote is active. A decision expires if voting is not completed within this period."
  - [x] Input feed: Admin can input the number of days for the voting duration
    - [x] Admin can only enter a number
    - [x] Default value is 7 days
  - [x] Button
    - [x] Cancel: Cancels any changes and leaves the voting duration unchanged
    - [x] Submit Requests: Confirm the changes and submits the request

## 2. Warning Message for Impact on Existing Requests 
- [x] Admin sees a warning message if the voting duration changes will impact any existing requests. See mock up below.

## 3. Submit Request to Update Voting Duration
- [x] When admin clicks Submit Request, a new policy change request is added to the "Pending Requests" section of the Settings page.
- [x] The requests will need to be voted on before the change is implemented. The voting thresholds for this policy change will be follow the "Manage Members & Settings" permission group. So if 1 vote is required, after 1 vote that policy change will take effect.
- [x] Once the votes are cast, the new voting duration policy will take effect.

## 4. Handling of Existing Requests
- [x] Once policy change takes effect, all existing requests that had the old vote duration policy will inherit the new voting duration policy:
  - [x] **Pending Requests:** All pending requests that were created under the previous duration will inherit the new voting period
  - [ ] **Expired Requests:** Any requests that were expired under the old voting duration but fall under the new voting duration will automatically move to the Pending Requests tab and be available for voting again

# Voting Duration Mockup
Image

# Voting Duration Warning Mockup
@FREZZZBE will add once https://github.com/NEAR-DevHub/neardevhub-treasury-dashboard/issues/119 is complete.

### LINKED ISSUE #127: Testing: Create testing environment for sputnik-dao interactions without requiring spending real NEAR
Use near-workspaces to launch sputnik-dao in a sandbox. Also patch SocialVM  to redirect transactions to the sandbox RPC.

### PR DESCRIPTION
**Pull request by petersalomonsen**
https://github.com/user-attachments/assets/af591569-480e-4e9b-aa9e-16304380e33b

Lowering voting duration should warn and show proposals that will expire because of the change

https://github.com/user-attachments/assets/d5829905-0643-4f57-96fe-b4abebe52c60

When a voting duration change proposal is submitted, a toast will show:

https://github.com/user-attachments/assets/1f004b39-d66c-4b1d-aa4a-57d5ed2047b2

The last test here uses a RPC sandbox, interacting with a real instance of the smart contract. Not on the mainnet, but in a near-workspaces-rs sandbox. This new sandbox facilitates real interaction with the smart contracts in tests, without having to do real mainnet transactions. Also there is a "playground" script which you can use to check how a contract works. Currently only the sputnik-dao contract is deployed.

resolves #116
resolves #127

### CODE REVIEW

**Comment by petersalomonsen** [comment]
Check out the video that shows which proposals that will expire when lowering the duration days. @ori-near @FREZZZBE

**Comment by petersalomonsen** [comment]
@race-of-sloths include

**Comment by race-of-sloths** [comment]
@petersalomonsen Thank you for your contribution! Your pull request is now a part of the [Race of Sloths](https://race-of-sloths.com)!
Do you want to apply for monthly streak? Get 8+ score for a single PR this month and receive boost for race-of-sloths!

[<picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://race-of-sloths.com/images/bot-banner-dark.svg">
    <source media="(prefers-color-scheme: light)" srcset="https://race-of-sloths.com/images/bot-banner-light.svg">
    Shows inviting banner with latest news.
</picture>](https://race-of-sloths.com/latest-news)

[<picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://badge.race-of-sloths.com/petersalomonsen?type=bot&pr=NEAR-DevHub/neardevhub-treasury-dashboard/121&theme=dark">
    <source media="(prefers-color-scheme: light)" srcset="https://badge.race-of-sloths.com/petersalomonsen?type=bot&pr=NEAR-DevHub/neardevhub-treasury-dashboard/121&theme=light">
    Shows profile picture for the author of the PR
</picture>](https://race-of-sloths.com/profile/petersalomonsen)


<details><summary>Current status: <i>waiting for scoring</i></summary>

We're waiting for maintainer to score this pull request with `@race-of-sloths score [0,1,2,3,5,8,13]` command. Alternatively, autoscoring [1,2] will be applied for this pull request

</details><details><summary>What is the Race of Sloths</summary>
<p>
Race of Sloths is a friendly competition where you can participate in challenges and compete with other open-source contributors within your normal workflow

For contributors:

- Tag @race-of-sloths inside your pull requests
- Wait for the maintainer to review and score your pull request
- Check out your position in the [Leaderboard](https://race-of-sloths.com/leaderboard)
- Keep weekly and monthly streaks to reach higher positions
- Boast your contributions with a dynamic picture of your [Profile](https://race-of-sloths.com/profile/petersalomonsen)

For maintainers:
- Score pull requests that participate in the Race of Sloths and receive a reward
- Engage contributors with fair scoring and fast responses so they keep their streaks
- Promote the Race to the point where the Race starts promoting you
- Grow the community of your contributors

Feel free to check [our website](https://race-of-sloths.com) for additional details!
</p>
</details> 

<details>
<summary>Bot commands</summary>

- For contributors
  - **Include a PR:** `@race-of-sloths include` to enter the Race with your PR
- For maintainers:
  - **Invite contributor** `@race-of-sloths invite` to invite the contributor to participate in a race or include it, if it's already a runner.
  - **Assign points:** `@race-of-sloths score [1/2/3/5/8/13]` to award points based on your assessment.
  - **Reject this PR:** `@race-of-sloths exclude` to send this PR back to the drawing board.
  - **Exclude repo:** `@race-of-sloths pause` to stop bot activity in this repo until `@race-of-sloths unpause` command is called

</details>

File: instances/treasury-devdao.near/widget/pages/settings/VotingDurationPage.jsx
Comment (Megha-Dev-19) [inline instances/treasury-devdao.near/widget/pages/settings/VotingDurationPage.jsx:137]: Don't we need to send all [parameters](https://github.com/near-daos/sputnik-dao-contract/blob/main/sputnikdao2/src/proposals.rs#L115)?
And I think we will need to update permissions as well (for all instances), because if I remember correctly `ChangePolicyUpdateParameters` isn't mentioned in policy which is set currently.

File: instances/treasury-devdao.near/widget/pages/settings/VotingDurationPage.jsx
Comment (Megha-Dev-19) [inline instances/treasury-devdao.near/widget/pages/settings/VotingDurationPage.jsx:154]: can we remove the logs?

**Review by Megha-Dev-19 — APPROVED** [review]
Looks great @petersalomonsen, thanks for attaching the test videos, left some comments.

**Comment by Megha-Dev-19** [comment]
@race-of-sloths score 13

File: instances/treasury-devdao.near/widget/pages/settings/VotingDurationPage.jsx
Comment (petersalomonsen) [reply instances/treasury-devdao.near/widget/pages/settings/VotingDurationPage.jsx:137]: When I look at the implementation of `update_parameters` here: https://github.com/near-daos/sputnik-dao-contract/blob/main/sputnikdao2/src/policy.rs#L270

I understand it as only the parameters you specify are updated, and the others are not touched. So I guess we don't have to send all parameters..

Regarding permissions. The test pass for adding the proposal, and there's also the [sandboxrpcplayground.js](https://github.com/NEAR-DevHub/neardevhub-treasury-dashboard/blob/e3c87a1cbad44313bca7fb9165022d6983c91896/sandboxrpc/sandboxrpcplayground.js) where I just played with the sputnik-dao contract to see how it works. As you can see there I add the proposal for changing the duration, and also vote for it, and you can see it takes effect. It seems that it works fine with the permission set up in that policy there. Would you take a look at that playground file and see if it makes sense?

File: instances/treasury-devdao.near/widget/pages/settings/VotingDurationPage.jsx
Comment (petersalomonsen) [reply instances/treasury-devdao.near/widget/pages/settings/VotingDurationPage.jsx]: Thanks! Removed now

**Comment by petersalomonsen** [comment]
@Megha-Dev-19 This feature requires that we add the `policy_update_parameters:AddProposal` permission to a role. How/where can we add this permission to a new or existing role?

**Comment by petersalomonsen** [comment]
@Megha-Dev-19 I added `policy_update_parameters:AddProposal` to the `Manage Members` role and also succeeded in testing this new feature manually as you can see here: https://near.social/astraplusplus.ndctools.near/widget/home?tab=proposals&daoId=testing-astradao.sputnik-dao.near&page=dao

**Pull request merged**
